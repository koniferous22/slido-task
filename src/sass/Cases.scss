@use 'base' as base;

section#cases {
	display: flex;
	justify-content: center;
	align-items: center;
	
	div#cases-hint {
		flex-shrink: 2;
		@include base.disable-small-devices(600px);
	}
}