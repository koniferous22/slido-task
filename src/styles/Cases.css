section#cases {
	display: flex;
	justify-content: center;
	align-items: center;
	text-align: center;
}

section#cases div#cases-hint {
	flex-shrink: 2;
	margin: auto 0;
	@media screen and (max-width: 600px) {
		display: none;
	}
}