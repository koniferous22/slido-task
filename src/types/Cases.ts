import { Case } from 'types/common/Case'

export interface CasesProps {
	caseEntries: Array<Case>;
	handleSelectCase: (_: number) => void;
	selectedCaseIndex: number;
}
