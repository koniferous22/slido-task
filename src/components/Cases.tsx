import React from 'react'

import CaseCategory from './cases/CaseCategory'
import { Case } from 'types/common/Case'
import { CasesProps } from 'types/Cases'

import 'styles/Cases.css'

const CasesRenderer: React.FC<CasesProps> = ({caseEntries, selectedCaseIndex, handleSelectCase}: CasesProps) => {
	const shouldDisplayPast: boolean = selectedCaseIndex > 0
	const shouldDisplayFuture: boolean = selectedCaseIndex + 1 < caseEntries.length
	const shouldDisplayHint: boolean = caseEntries.length > 1

	const pastCases: Array<Case> = caseEntries.slice(0, selectedCaseIndex)
	const futureCases: Array<Case> = caseEntries.slice(selectedCaseIndex + 1)

	return (
		<section id="cases">
			{
				shouldDisplayPast && 
					<CaseCategory
						id="past"
						label="Past Events"
						caseEntries={pastCases}
						handleSelectCase={handleSelectCase}
					/>
			}
			{
				shouldDisplayHint &&
					<div id="cases-hint">
						<h4>Pick an event to display</h4>
						<h6>lists are scrollable :)</h6>
					</div>
			}
			{
				shouldDisplayFuture && 
					<CaseCategory 
						id="future"
						label="Future Events"
						caseEntries={futureCases}
						handleSelectCase={handleSelectCase} 
						offset={selectedCaseIndex + 1}
					/>
			}
		</section>
	)	
}

export default CasesRenderer
