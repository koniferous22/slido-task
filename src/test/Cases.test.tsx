import React from 'react'
import renderer from 'react-test-renderer'
import Cases from 'components/Cases'

import { cleanup, fireEvent, render } from '@testing-library/react'

import { shortenTitle } from 'utils/functions'

import { sampleCases } from 'test/data/sampleCases'

afterEach(cleanup)

const renderTests: Array<string> = ['future only', 'perfectly balanced', 'past only']

for (let selectedCase = 0; selectedCase < renderTests.length; ++selectedCase) {
	it('CaseAddForm renders correctly: ' + renderTests[selectedCase], () => {
		const tree = renderer
			.create(<Cases
				caseEntries={sampleCases}
				selectedCaseIndex={selectedCase}
				handleSelectCase={(): void => {return;}}
			/>)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
}

it('Click past case', () => {
	const mockSelectCase = jest.fn();	
	const { getByText } = render(<Cases
		caseEntries={sampleCases}
		selectedCaseIndex={1}
		handleSelectCase={mockSelectCase}
	/>)
	fireEvent.click(getByText(shortenTitle(sampleCases[0].title, 50)));
	expect(mockSelectCase).toHaveBeenCalled()
})

it('Click future case', () => {
	const mockSelectCase = jest.fn();	
	const { getByText } = render(<Cases
		caseEntries={sampleCases}
		selectedCaseIndex={1}
		handleSelectCase={mockSelectCase}
	/>)
	fireEvent.click(getByText(shortenTitle(sampleCases[2].title, 50)));
	expect(mockSelectCase).toHaveBeenCalled()
})
