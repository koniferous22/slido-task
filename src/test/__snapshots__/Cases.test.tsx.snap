// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseAddForm renders correctly: future only 1`] = `
<section
  id="cases"
>
  <div
    id="cases-hint"
  >
    <h4>
      Pick an event to display
    </h4>
    <h6>
      lists are scrollable :)
    </h6>
  </div>
  <section
    id="future"
  >
    <h4
      className="case-list-title"
    >
      Future Events
    </h4>
    <ul
      className="case-list"
    >
      <li
        onClick={[Function]}
      >
        <i>
          [2/26/2020] 
        </i>
        Filakovsky jarmok
      </li>
      <li
        onClick={[Function]}
      >
        <i>
          [2/26/2020] 
        </i>
        I'm Rick Harrison, and this is my pawn shop.I work...
      </li>
    </ul>
  </section>
</section>
`;

exports[`CaseAddForm renders correctly: past only 1`] = `
<section
  id="cases"
>
  <section
    id="past"
  >
    <h4
      className="case-list-title"
    >
      Past Events
    </h4>
    <ul
      className="case-list"
    >
      <li
        onClick={[Function]}
      >
        <i>
          [1/26/2020] 
        </i>
        Dni Zelá v Stupave
      </li>
      <li
        onClick={[Function]}
      >
        <i>
          [2/26/2020] 
        </i>
        Filakovsky jarmok
      </li>
    </ul>
  </section>
  <div
    id="cases-hint"
  >
    <h4>
      Pick an event to display
    </h4>
    <h6>
      lists are scrollable :)
    </h6>
  </div>
</section>
`;

exports[`CaseAddForm renders correctly: perfectly balanced 1`] = `
<section
  id="cases"
>
  <section
    id="past"
  >
    <h4
      className="case-list-title"
    >
      Past Events
    </h4>
    <ul
      className="case-list"
    >
      <li
        onClick={[Function]}
      >
        <i>
          [1/26/2020] 
        </i>
        Dni Zelá v Stupave
      </li>
    </ul>
  </section>
  <div
    id="cases-hint"
  >
    <h4>
      Pick an event to display
    </h4>
    <h6>
      lists are scrollable :)
    </h6>
  </div>
  <section
    id="future"
  >
    <h4
      className="case-list-title"
    >
      Future Events
    </h4>
    <ul
      className="case-list"
    >
      <li
        onClick={[Function]}
      >
        <i>
          [2/26/2020] 
        </i>
        I'm Rick Harrison, and this is my pawn shop.I work...
      </li>
    </ul>
  </section>
</section>
`;
